pgjdbc/src/main/java/org/postgresql/util/LruCache.java

```diff
@@ -134,6 +180,13 @@ public synchronized Value borrow(Key key) throws SQLException {
    * @param value value
    */
   public synchronized void put(Key key, Value value) {
+    innerPut(key, value);
+  }
+
+  /**
+   * Actual work of put, but requires calling method to obtain lock.
+   */
+  private void innerPut(Key key, Value value) {
```

Comment:
Why do you need to do this? "synchronized" is already re-entrant. The prior code worked fine?

For all of these - please prove me wrong by providing a benchmark. It doesn't even have to be something that is absolutely real. Just a use case where calling putAll() was slow, and deserves adding this level of complexity to try to avoid re-entrant locks. I think you will be surprised, and find out that it doesn't really make a difference for real-world use cases. Java optimizes for re-entrant synchronized already, and you aren't saving much of anything.

Reply:
This was just an oversight as I was unwinding the stamped lock change.